CI failures - KafkaCodecTest is consistently failing

The `KafkaCodecTest` has started failing consistently in CI runs (JVM version doesn't seem to matter, since it has failed in Java 8, 11 and 14 from some of the runs I saw). The exception is pasted below, taken from one unrelated PR[1].

I can't reproduce this locally and looking at the code and test in question, there are a couple of issues I think:

1. This `KafkaSASLTestResource` (which is a `QuarkusTestResourceLifecycleManager`) shouldn't even be run for this `KafkaCodecTest` since it doesn't explicitly list it through `@QuarkusTestResource`. But having seen similar reports, I think this is a "known" thing and the test framework uses all available `QuarkusTestResourceLifecycleManager` in the classpath, irrespective of which one is a actually marked for use in the test?

2. I can't seen anything wrong that the `KafkaSASLTestResource` does. Digging deeper in that code, Kafka internally uses `javax.security.auth.login.Configuration.getConfiguration` for loading the JAAS configuration[2]. That method uses a system level configuration and as per the javadoc of that method, it returns back whatever was loaded or set previously during the lifetime of the JVM. What that hints to me is that something in some previous tests can end up setting/creating a `Configuration` which doesn't have this JAAS configuration that was written out in a temporary configuration file by this `KafkaSASLTestResource`, which then results in this slightly misleading error (since it seems to indicate that it indeed used the file that was set as the `java.security.auth.login.config` system property, but in reality, it probably just used a previously created `Configuration` instance).

I can't think of any easy ways around this and I suspect this might even need a fix in Kafka to allow cases like these? Other way perhaps is to unset and reset the `Configuration` instance within this `KafkaSASLTestResource`'s lifecycle, but then again I don't know what other impact it might have.

All being said, I don't know why this has started showing up now (I couldn't find any related changes in that test that could have led to this). So maybe something else has started contributing to this?



```
2020-07-03T20:40:23.7998358Z [INFO] -------------------------------------------------------
2020-07-03T20:40:23.7998750Z [INFO]  T E S T S
2020-07-03T20:40:23.7999345Z [INFO] -------------------------------------------------------
2020-07-03T20:40:24.8040241Z [INFO] Running io.quarkus.it.kafka.KafkaCodecTest
2020-07-03T20:40:58.6672640Z [ERROR] Tests run: 3, Failures: 0, Errors: 1, Skipped: 2, Time elapsed: 33.85 s <<< FAILURE! - in io.quarkus.it.kafka.KafkaCodecTest
2020-07-03T20:40:58.6682771Z [ERROR] io.quarkus.it.kafka.KafkaCodecTest.testJacksonCodec  Time elapsed: 0.014 s  <<< ERROR!
2020-07-03T20:40:58.6688059Z java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
2020-07-03T20:40:58.6695079Z 	at io.quarkus.test.junit.QuarkusTestExtension.throwBootFailureException(QuarkusTestExtension.java:424)
2020-07-03T20:40:58.6701359Z 	at io.quarkus.test.junit.QuarkusTestExtension.beforeEach(QuarkusTestExtension.java:337)
2020-07-03T20:40:58.6710146Z 	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeBeforeEachCallbacks$1(TestMethodTestDescriptor.java:161)
2020-07-03T20:40:58.6728795Z 	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeBeforeMethodsOrCallbacksUntilExceptionOccurs$5(TestMethodTestDescriptor.java:197)
2020-07-03T20:40:58.6736193Z 	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
2020-07-03T20:40:58.6754486Z 	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeBeforeMethodsOrCallbacksUntilExceptionOccurs(TestMethodTestDescriptor.java:197)
2020-07-03T20:40:58.6779197Z 	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeBeforeEachCallbacks(TestMethodTestDescriptor.java:160)
2020-07-03T20:40:58.6786614Z 	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:131)
2020-07-03T20:40:58.6794251Z 	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:71)
2020-07-03T20:40:58.6802234Z 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:135)
2020-07-03T20:40:58.6814582Z 	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
2020-07-03T20:40:58.6833342Z 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:125)
2020-07-03T20:40:58.6841786Z 	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:135)
2020-07-03T20:40:58.6849478Z 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:123)
2020-07-03T20:40:58.6861153Z 	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
2020-07-03T20:40:58.6869617Z 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:122)
2020-07-03T20:40:58.6880175Z 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:80)
2020-07-03T20:40:58.7060968Z 	at java.util.ArrayList.forEach(ArrayList.java:1257)
2020-07-03T20:40:58.7063718Z 	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:38)
2020-07-03T20:40:58.7065287Z 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:139)
2020-07-03T20:40:58.7066374Z 	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
2020-07-03T20:40:58.7077333Z 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:125)
2020-07-03T20:40:58.7078572Z 	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:135)
2020-07-03T20:40:58.7079657Z 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:123)
2020-07-03T20:40:58.7080735Z 	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
2020-07-03T20:40:58.7081812Z 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:122)
2020-07-03T20:40:58.7091434Z 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:80)
2020-07-03T20:40:58.7092584Z 	at java.util.ArrayList.forEach(ArrayList.java:1257)
2020-07-03T20:40:58.7093687Z 	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:38)
2020-07-03T20:40:58.7095074Z 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:139)
2020-07-03T20:40:58.7096181Z 	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
2020-07-03T20:40:58.7097265Z 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:125)
2020-07-03T20:40:58.7136184Z 	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:135)
2020-07-03T20:40:58.7137388Z 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:123)
2020-07-03T20:40:58.7138491Z 	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
2020-07-03T20:40:58.7139578Z 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:122)
2020-07-03T20:40:58.7140657Z 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:80)
2020-07-03T20:40:58.7141778Z 	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:32)
2020-07-03T20:40:58.7142870Z 	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
2020-07-03T20:40:58.7143940Z 	at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:51)
2020-07-03T20:40:58.7145009Z 	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:248)
2020-07-03T20:40:58.7146102Z 	at org.junit.platform.launcher.core.DefaultLauncher.lambda$execute$5(DefaultLauncher.java:211)
2020-07-03T20:40:58.7147183Z 	at org.junit.platform.launcher.core.DefaultLauncher.withInterceptedStreams(DefaultLauncher.java:226)
2020-07-03T20:40:58.7148259Z 	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:199)
2020-07-03T20:40:58.7205327Z 	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:132)
2020-07-03T20:40:58.7213034Z 	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.execute(JUnitPlatformProvider.java:188)
2020-07-03T20:40:58.7235320Z 	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.invokeAllTests(JUnitPlatformProvider.java:154)
2020-07-03T20:40:58.7242652Z 	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.invoke(JUnitPlatformProvider.java:128)
2020-07-03T20:40:58.7263262Z 	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:428)
2020-07-03T20:40:58.7273445Z 	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:162)
2020-07-03T20:40:58.7307303Z 	at org.apache.maven.surefire.booter.ForkedBooter.run(ForkedBooter.java:562)
2020-07-03T20:40:58.7312943Z 	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:548)
2020-07-03T20:40:58.7317596Z Caused by: java.lang.reflect.InvocationTargetException
2020-07-03T20:40:58.7322438Z 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
2020-07-03T20:40:58.7328206Z 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
2020-07-03T20:40:58.7334561Z 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
2020-07-03T20:40:58.7346779Z 	at java.lang.reflect.Method.invoke(Method.java:498)
2020-07-03T20:40:58.7355946Z 	at io.quarkus.test.junit.QuarkusTestExtension.doJavaStart(QuarkusTestExtension.java:197)
2020-07-03T20:40:58.7364639Z 	at io.quarkus.test.junit.QuarkusTestExtension.ensureStarted(QuarkusTestExtension.java:402)
2020-07-03T20:40:58.7375858Z 	at io.quarkus.test.junit.QuarkusTestExtension.beforeAll(QuarkusTestExtension.java:435)
2020-07-03T20:40:58.7393872Z 	at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.lambda$invokeBeforeAllCallbacks$7(ClassBasedTestDescriptor.java:359)
2020-07-03T20:40:58.7401339Z 	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
2020-07-03T20:40:58.7424380Z 	at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.invokeBeforeAllCallbacks(ClassBasedTestDescriptor.java:359)
2020-07-03T20:40:58.7432004Z 	at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.before(ClassBasedTestDescriptor.java:189)
2020-07-03T20:40:58.7439361Z 	at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.before(ClassBasedTestDescriptor.java:78)
2020-07-03T20:40:58.7511555Z 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:132)
2020-07-03T20:40:58.7513576Z 	... 32 more
2020-07-03T20:40:58.7527571Z Caused by: java.lang.RuntimeException: Unable to start Quarkus test resource io.quarkus.it.kafka.KafkaSASLTestResource@41059616
2020-07-03T20:40:58.7533467Z 	at io.quarkus.test.common.TestResourceManager.start(TestResourceManager.java:57)
2020-07-03T20:40:58.7535294Z 	... 45 more
2020-07-03T20:40:58.7556563Z Caused by: java.lang.RuntimeException: java.lang.IllegalArgumentException: Could not find a 'KafkaServer' or 'client.KafkaServer' entry in the JAAS configuration. System property 'java.security.auth.login.config' is /home/runner/work/quarkus/quarkus/integration-tests/kafka/target/kafka-server-jaas.6477489435657081630.conf
2020-07-03T20:40:58.7562826Z 	at io.quarkus.it.kafka.KafkaSASLTestResource.start(KafkaSASLTestResource.java:46)
2020-07-03T20:40:58.7568602Z 	at io.quarkus.test.common.TestResourceManager.start(TestResourceManager.java:52)
2020-07-03T20:40:58.7570439Z 	... 45 more
2020-07-03T20:40:58.7594157Z Caused by: java.lang.IllegalArgumentException: Could not find a 'KafkaServer' or 'client.KafkaServer' entry in the JAAS configuration. System property 'java.security.auth.login.config' is /home/runner/work/quarkus/quarkus/integration-tests/kafka/target/kafka-server-jaas.6477489435657081630.conf
2020-07-03T20:40:58.7602219Z 	at org.apache.kafka.common.security.JaasContext.defaultContext(JaasContext.java:133)
2020-07-03T20:40:58.7608041Z 	at org.apache.kafka.common.security.JaasContext.load(JaasContext.java:98)
2020-07-03T20:40:58.7614238Z 	at org.apache.kafka.common.security.JaasContext.loadServerContext(JaasContext.java:70)
2020-07-03T20:40:58.7620500Z 	at org.apache.kafka.common.network.ChannelBuilders.create(ChannelBuilders.java:131)
2020-07-03T20:40:58.7627305Z 	at org.apache.kafka.common.network.ChannelBuilders.serverChannelBuilder(ChannelBuilders.java:97)
2020-07-03T20:40:58.7631779Z 	at kafka.network.Processor.<init>(SocketServer.scala:724)
2020-07-03T20:40:58.7638376Z 	at kafka.network.SocketServer.newProcessor(SocketServer.scala:367)
2020-07-03T20:40:58.7644567Z 	at kafka.network.SocketServer.$anonfun$addDataPlaneProcessors$1(SocketServer.scala:252)
2020-07-03T20:40:58.7651899Z 	at kafka.network.SocketServer.addDataPlaneProcessors(SocketServer.scala:251)
2020-07-03T20:40:58.7659348Z 	at kafka.network.SocketServer.$anonfun$createDataPlaneAcceptorsAndProcessors$1(SocketServer.scala:214)
2020-07-03T20:40:58.7667075Z 	at kafka.network.SocketServer.$anonfun$createDataPlaneAcceptorsAndProcessors$1$adapted(SocketServer.scala:211)
2020-07-03T20:40:58.7673229Z 	at scala.collection.mutable.ResizableArray.foreach(ResizableArray.scala:62)
2020-07-03T20:40:58.7682561Z 	at scala.collection.mutable.ResizableArray.foreach$(ResizableArray.scala:55)
2020-07-03T20:40:58.7687821Z 	at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:49)
2020-07-03T20:40:58.7723066Z 	at kafka.network.SocketServer.createDataPlaneAcceptorsAndProcessors(SocketServer.scala:211)
2020-07-03T20:40:58.7727871Z 	at kafka.network.SocketServer.startup(SocketServer.scala:122)
2020-07-03T20:40:58.7732393Z 	at kafka.server.KafkaServer.startup(KafkaServer.scala:266)
2020-07-03T20:40:58.7747074Z 	at io.debezium.kafka.KafkaServer.startup(KafkaServer.java:220)
2020-07-03T20:40:58.7748241Z 	at java.util.concurrent.ConcurrentHashMap$ValuesView.forEach(ConcurrentHashMap.java:4707)
2020-07-03T20:40:58.7749333Z 	at io.debezium.kafka.KafkaCluster.startup(KafkaCluster.java:250)
2020-07-03T20:40:58.7750395Z 	at io.quarkus.it.kafka.KafkaSASLTestResource.start(KafkaSASLTestResource.java:44)
2020-07-03T20:40:58.7751432Z 	... 46 more
2020-07-03T20:40:58.7752411Z 
```

[1] https://github.com/quarkusio/quarkus/pull/10464/checks?check_run_id=835445764
[2] https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/common/security/JaasContext.java#L114

--- a/integration-tests/kafka/src/test/java/io/quarkus/it/kafka/KafkaSASLTestResource.java
+++ b/integration-tests/kafka/src/test/java/io/quarkus/it/kafka/KafkaSASLTestResource.java
@@ -1,14 +1,11 @@
 package io.quarkus.it.kafka;
 
 import java.io.File;
-import java.io.IOException;
-import java.nio.charset.StandardCharsets;
-import java.nio.file.Files;
-import java.nio.file.Path;
-import java.nio.file.StandardOpenOption;
 import java.util.Collections;
 import java.util.Map;
 import java.util.Properties;
+
+import org.apache.kafka.common.config.SaslConfigs;
 
 import io.debezium.kafka.KafkaCluster;
 import io.debezium.util.Testing;
@@ -23,8 +20,6 @@ public class KafkaSASLTestResource implements QuarkusTestResourceLifecycleManage
         try {
             File directory = Testing.Files.createTestingDirectory("kafka-data-sasl", true);
 
-            enableServerJaasConf();
-
             Properties props = new Properties();
             props.setProperty("zookeeper.connection.timeout.ms", "45000");
             props.setProperty("listener.security.protocol.map", "CLIENT:SASL_PLAINTEXT");
@@ -33,6 +28,11 @@ public class KafkaSASLTestResource implements QuarkusTestResourceLifecycleManage
 
             props.setProperty("sasl.enabled.mechanisms", "PLAIN");
             props.setProperty("sasl.mechanism.inter.broker.protocol", "PLAIN");
+
+            final String jaasConf = "org.apache.kafka.common.security.plain.PlainLoginModule required" +
+                    " username=broker password=broker-secret" +
+                    " user_broker=broker-secret user_client=client-secret;";
+            props.setProperty("listener.name.client.plain." + SaslConfigs.SASL_JAAS_CONFIG, jaasConf);
 
             kafka = new KafkaCluster()
                     .withPorts(2184, 19094)
@@ -54,34 +54,6 @@ public class KafkaSASLTestResource implements QuarkusTestResourceLifecycleManage
         if (kafka != null) {
             kafka.shutdown();
         }
-    }
-
-    public static void enableServerJaasConf() throws IOException {
-        final Path conf = Files.createTempFile("kafka-server-jaas.", ".conf");
-        String serverConfiguration = "KafkaServer { "
-                + "org.apache.kafka.common.security.plain.PlainLoginModule required "
-                + "username=\"broker\" "
-                + "password=\"broker-secret\" "
-                + "user_broker=\"broker-secret\" "
-                + "user_client=\"client-secret\"; };";
-
-        Files.write(conf, ("client." + serverConfiguration).getBytes(StandardCharsets.UTF_8),
-                StandardOpenOption.CREATE);
-
-        System.setProperty("java.security.auth.login.config", conf.toAbsolutePath().toString());
-        System.setProperty("zookeeper.sasl.client", "false");
-    }
-
-    /**
-     * Make sure this runs first otherwise system property {@code java.security.auth.login.config}
-     * is ignored since {@link javax.security.auth.login.Configuration#configuration}
-     * is already initialized with default, empty configuration.
-     *
-     * @return {@link Integer#MIN_VALUE}
-     */
-    @Override
-    public int order() {
-        return Integer.MIN_VALUE;
     }
 
 }

